### ISSUE TITLE
Profiling zeebe for memory usage

### ISSUE THREAD

**Issue by deepthidevaki** [zeebe-io/zeebe#5736]
**Description**

Based on the result from this issue  https://github.com/zeebe-io/zeebe/issues/3988 we configured the following parameters to reduce the memory footprint of Zeebe. 

ZEEBE_BROKER_DATA_ROCKSDB_COLUMNFAMILYOPTIONS_WRITE_BUFFER_SIZE=8MB
ZEEBE_BROKER_DATA_ROCKSDB_COLUMNFAMILYOPTIONS_MAX_WRITE_BUFFER_SIZE_TO_MAINTAIN=8MB

We had also added rocksdb metrics so we can keep track of rocksdb memory usage.

As a result we observed that rocksdb memory usage reported by the metrics is capped. However we observe that the total memory consumed by the process as reported by the metrics `container_memory_rss` is slowly increasing. Some users also reported that the broker was killed due to OOM after a while, eventhough reported rocksdb memory + java heap + non-heap is under the limit.

In this issue, we want to investigate what is actually consuming the missing memory and what is causing OOM?

**Comment by deepthidevaki** [comment]
To profile native memory usage, I replaced the default memory allocator `glibc` with other libraries. I used both [tcmalloc](https://gperftools.github.io/gperftools/heapprofile.html)
and [jemalloc](https://github.com/jemalloc/jemalloc/wiki/).

`tcmalloc` instruments all calls to malloc and free. So the brokers became too slow.
`jemalloc` only samples, but it gives similar insight as `tcmalloc`.

To profile zeebe cluster I build a new docker image with following libraries installed. ` binutils, libjemalloc2, libgoogle-perftools4, google-perftools`

To use tcmalloc with profiling enabled set the following environment variables.

```
export HEAPPROFILE=/usr/local/zeebe/data/pprof
export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_and_profiler.so.4 bin/broker
```

**Comment by deepthidevaki** [comment]
Running a cluster with clusterSize: 3, partitionCount:3 and replicationFactor:3. At first I profiled it with tcmalloc.

Following is the log from one of the broker - leader for one partition. Notice the heap currently in use
```

E 2020-10-28T12:08:47.927665613Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0039.heap (38371 MB allocated cumulatively, 496 MB currently in use) 
E 2020-10-28T12:10:41.352281896Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0040.heap (39395 MB allocated cumulatively, 512 MB currently in use) 
E 2020-10-28T12:12:33.715094387Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0041.heap (40419 MB allocated cumulatively, 515 MB currently in use)
… 
 
E 2020-10-28T14:08:24.360030721Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0102.heap (102884 MB allocated cumulatively, 629 MB currently in use)
E 2020-10-28T14:10:18.199275732Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0103.heap (103908 MB allocated cumulatively, 620 MB currently in use) 
E 2020-10-28T14:12:15.360656082Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0104.heap (104933 MB allocated cumulatively, 631 MB currently in use)
```
![image](https://user-images.githubusercontent.com/1997478/97724811-b1e31800-1acd-11eb-9710-73425b0b5c57.png)
![image](https://user-images.githubusercontent.com/1997478/97724833-b8718f80-1acd-11eb-92de-84667db62488.png)
![image](https://user-images.githubusercontent.com/1997478/97724854-be677080-1acd-11eb-94fc-51f2efc5f3fa.png)


Following is the output from 
google-pprof --text --base zeebe.hprof_6.0040.heap `which java` zeebe.hprof_6.0103.heap
It indicates within 2 hours an additional 107MB is added to the heap out of which 104 is coming from `rocksdb::Arena::AllocateNewBlock`
```
Total: 107.8 MB
   104.0  96.5%  96.5%    104.0  96.5% rocksdb::Arena::AllocateNewBlock
     3.2   3.0%  99.5%      3.2   3.0% os::malloc@bb9170
     0.3   0.3%  99.8%      0.5   0.4% rocksdb::ColumnFamilyData::ConstructNewMemtable
     0.1   0.1%  99.9%      0.1   0.1% rocksdb::ConcurrentArena::ConcurrentArena
     0.1   0.1% 100.0%      0.1   0.1% rocksdb::WriteBatchWithIndex::WriteBatchWithIndex
     0.0   0.0% 100.0%    104.0  96.5% rocksdb::SkipListFactory::CreateMemTableRep
     0.0   0.0% 100.0%      0.0   0.0% std::vector::_M_realloc_insert
     0.0   0.0% 100.0%      0.0   0.0% std::basic_string::_Rep::_S_create
     0.0   0.0% 100.0%      0.0   0.0% rocksdb::UncompressBlockContentsForCompressionType
     0.0   0.0% 100.0%      0.0   0.0% rocksdb::WalManager::ReadFirstRecord
     0.0   0.0% 100.0%      0.1   0.1% rocksdb::OptimisticTransactionDBImpl::BeginTransaction
     0.0   0.0% 100.0%      0.0   0.0% std::_Hashtable::_M_rehash
     0.0   0.0% 100.0%      0.0   0.0% rocksdb::ConcurrentArena::Allocate
     0.0   0.0% 100.0%      0.0   0.0% std::_Rb_tree::_M_insert_unique@48f990
     0.0   0.0% 100.0%      0.0   0.0% std::_Rb_tree::_M_insert_unique@4daf30
     0.0   0.0% 100.0%      0.0   0.0% rocksdb::BlockBasedTable::Open
     0.0   0.0% 100.1%      0.0   0.0% rocksdb::ReadProperties
     0.0   0.0% 100.1%      0.0   0.0% rocksdb::VersionBuilder::Rep::Apply
     0.0   0.0% 100.1%      0.0   0.0% rocksdb::TableCache::GetTableReader
     0.0   0.0% 100.1%    110.1 102.2% 0x00007fb7c3e48848

 <output trimmed>
 ```

However Rocksdb metrics shows total memory usage stays same.

**Comment by deepthidevaki** [comment]
The cluster was left running overnight. Heap usage is not increasing anymore. It is stabilized around 820 MB for more than 12 hours
```
E 2020-10-28T18:18:38.421253851Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0223.heap (226791 MB allocated cumulatively, 822 MB currently in use) 
E 2020-10-28T18:20:44.476469956Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0224.heap (227815 MB allocated cumulatively, 830 MB currently in use) 
E 2020-10-28T18:22:41.782859380Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0225.heap (228839 MB allocated cumulatively, 831 MB currently in use)
…

E 2020-10-29T08:33:48.175397011Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0656.heap (670191 MB allocated cumulatively, 805 MB currently in use) 
E 2020-10-29T08:35:46.245874295Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0657.heap (671215 MB allocated cumulatively, 821 MB currently in use) 
E 2020-10-29T08:37:43.237213652Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0658.heap (672239 MB allocated cumulatively, 824 MB currently in use)
E 2020-10-29T08:39:35.227599307Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0659.heap (673263 MB allocated cumulatively, 811 MB currently in use) 
E 2020-10-29T08:41:36.256830392Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0660.heap (674287 MB allocated cumulatively, 824 MB currently in use) 
E 2020-10-29T08:43:33.159840790Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0661.heap (675311 MB allocated cumulatively, 831 MB currently in use) 
E 2020-10-29T08:45:22.753197351Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0662.heap (676335 MB allocated cumulatively, 817 MB currently in use) 
E 2020-10-29T08:47:16.903854165Z Dumping heap profile to /usr/local/zeebe/data/zeebe.hprof_6.0663.heap (677359 MB allocated cumulatively, 820 MB currently in use)
```
However RSS is still increasing slowly. Note that RSS increase may not be harmful, as it also contains memory mapped. And the OS may not release memory unless it is under memory pressure.

The pod was restarted because of node preemptions. So we don't know if it goes out of memory or not.

**Comment by deepthidevaki** [comment]
Profiling with tcmalloc makes the system too slow. Instead we can also use jemalloc. It would give similar insights as tcmalloc.
To use jemalloc with profiling set the following environment variables:
```
export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2
export MALLOC_CONF=prof:true,lg_prof_interval:30,lg_prof_sample:17,prof_prefix:/usr/local/zeebe/data/jeprof
```
And use `jeprof -pdf `which java` jeprof.24301.10.i10.heap > prof.pdf` to anlayze the results.

**Comment by deepthidevaki** [comment]
In an attempt to reproduce OOM, I started a cluster with following configuration:

```
clusterSize: 1
partitionCount: 3
replicationFactor: 1

- name: ZEEBE_BROKER_DATA_ROCKSDB_COLUMNFAMILYOPTIONS_WRITE_BUFFER_SIZE
    value: 8MB
  - name: ZEEBE_BROKER_DATA_ROCKSDB_COLUMNFAMILYOPTIONS_MAX_WRITE_BUFFER_SIZE_TO_MAINTAIN
    value: 8MB
  - name: ZEEBE_BROKER_DATA_USEMMAP
    value: "true"
  - name: ZEEBE_BROKER_DATA_SNAPSHOTPERIOD
    value: "1m"
```
and workload consisting of a single task workflow and a timer workflow (available in our benchmark module). Also added a script that publish a message every second, even though there are no workflows with messages. 
In this cluster, we can observe that memory used is increasing much faster than usual. The cluster is running for more than 20hrs now. jemalloc profiling results shows that memory allocated by rocksdb is increasing, is not yet capped.

![image](https://user-images.githubusercontent.com/1997478/98082538-32609a80-1e79-11eb-8d54-9418be83cf8f.png)

The difference in memory allocation in 1hr is as follows: (comparing two heapdumps which are 1hr apart).
![image](https://user-images.githubusercontent.com/1997478/98082639-5d4aee80-1e79-11eb-8b78-b941e2508be1.png)

However RocksDB metrics reports a stable memory usage, which is capped at aroung 1.5g
![image](https://user-images.githubusercontent.com/1997478/98082778-94b99b00-1e79-11eb-8c65-168469723427.png)

If we leave the cluster running, there is a good chance that it will go OOM.
![image](https://user-images.githubusercontent.com/1997478/98084000-79e82600-1e7b-11eb-9962-db44f8b55371.png)

**Comment by deepthidevaki** [comment]
Taking snapshot too frequently can lead to high memory usage.

I have been running a two cluster with same configuration - except one with snapshot period = 1m and the other with snapshot period = 15m. 

Results for snapshotPeriod=15m:

From old to new
At 02:26
```
Total: 1276.0 MB
   579.0  45.4%  45.4%    579.0  45.4% rocksdb::Arena::AllocateNewBlock
   484.8  38.0%  83.4%    484.8  38.0% os::malloc@bb9170
   145.7  11.4%  94.8%    174.5  13.7% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.9%  96.7%     24.0   1.9% std::vector::reserve
```

```
Total: 1328.9 MB
   628.0  47.3%  47.3%    628.0  47.3% rocksdb::Arena::AllocateNewBlock
   487.7  36.7%  84.0%    487.7  36.7% os::malloc@bb9170
   145.7  11.0%  94.9%    174.5  13.1% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.8%  96.7%     24.0   1.8% std::vector::reserve
```

```
Total: 1303.1 MB
   602.0  46.2%  46.2%    602.0  46.2% rocksdb::Arena::AllocateNewBlock
   488.3  37.5%  83.7%    488.3  37.5% os::malloc@bb9170
   145.7  11.2%  94.9%    174.5  13.4% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.8%  96.7%     24.0   1.8% std::vector::reserve
```
At 11:45
```
Total: 1274.7 MB
   566.0  44.4%  44.4%    566.0  44.4% rocksdb::Arena::AllocateNewBlock
   496.4  38.9%  83.3%    496.4  38.9% os::malloc@bb9170
   145.7  11.4%  94.8%    174.5  13.7% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.9%  96.7%     24.0   1.9% std::vector::reserve
```
![image](https://user-images.githubusercontent.com/1997478/98234578-f3a60f80-1f60-11eb-95ca-835f1c925407.png)

Results for snapshotPeriod=1m:

At 01:18
```
Total: 2241.8 MB
  1532.0  68.3%  68.3%   1532.0  68.3% rocksdb::Arena::AllocateNewBlock
   489.3  21.8%  90.2%    489.3  21.8% os::malloc@bb9170
   145.1   6.5%  96.6%    173.7   7.7% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.1%  97.7%     24.0   1.1% std::vector::reserve
```

```
Total: 2296.1 MB
  1583.0  68.9%  68.9%   1583.0  68.9% rocksdb::Arena::AllocateNewBlock
   490.9  21.4%  90.3%    490.9  21.4% os::malloc@bb9170
   145.1   6.3%  96.6%    173.7   7.6% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.0%  97.7%     24.0   1.0% std::vector::reserve
```

```
Total: 2351.9 MB
  1636.0  69.6%  69.6%   1636.0  69.6% rocksdb::Arena::AllocateNewBlock
   491.7  20.9%  90.5%    491.7  20.9% os::malloc@bb9170
   145.1   6.2%  96.6%    173.7   7.4% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.0%  97.7%     24.0   1.0% std::vector::reserve
```

At 11:45
```
Total: 2417.7 MB
  1696.0  70.1%  70.1%   1696.1  70.2% rocksdb::Arena::AllocateNewBlock
   493.0  20.4%  90.5%    493.0  20.4% os::malloc@bb9170
   145.1   6.0%  96.5%    173.7   7.2% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.0%  97.5%     24.0   1.0% std::vector::reserve
```
![image](https://user-images.githubusercontent.com/1997478/98234516-de30e580-1f60-11eb-8fce-35c6e0343eff.png)

**Comment by deepthidevaki** [comment]
Also running a benchmark with the following additional rocksdb config. These configs are not exposed, so had to change the code.

```
final LRUCache cache = new LRUCache(64 * 1024 * 1024L, 6, false);
final WriteBufferManager writeBufferManager = new WriteBufferManager(256 * 1024 * 1024L, cache);
final BlockBasedTableConfig tableConfig =
          new BlockBasedTableConfig()
              .setBlockCache(cache)
              .setCacheIndexAndFilterBlocks(true)
              .setOptimizeFiltersForMemory(true);
columnFamilyOptions.setTableFormatConfig(tableConfig);
dbOptions.
              .setWriteBufferManager(writeBufferManager)
              .setMaxTotalWalSize(1024 * 1024 * 1024);
```

At 00:00
```
Total: 2467.4 MB
  1777.0  72.0%  72.0%   1778.3  72.1% rocksdb::Arena::AllocateNewBlock
   487.0  19.7%  91.8%    487.0  19.7% os::malloc@bb9170
   147.5   6.0%  97.7%    175.6   7.1% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.0%  98.7%     24.0   1.0% std::vector::reserve
```

At 12:04
```
Total: 2534.0 MB
  1831.0  72.3%  72.3%   1832.3  72.3% rocksdb::Arena::AllocateNewBlock
   491.0  19.4%  91.6%    491.0  19.4% os::malloc@bb9170
   147.5   5.8%  97.5%    175.6   6.9% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   0.9%  98.4%     24.0   0.9% std::vector::reserve
```
![image](https://user-images.githubusercontent.com/1997478/98239294-feb06e00-1f67-11eb-93eb-60fe0884b98a.png)

**Comment by npepinpe** [comment]
Well this isn't behaving at all as expected :sweat_smile: I thought write buffer manager would limit the total memtable size across the whole DB to 256Mb...did I misunderstand the setting?

**Comment by deepthidevaki** [comment]
> Well this isn't behaving at all as expected sweat_smile I thought write buffer manager would limit the total memtable size across the whole DB to 256Mb...did I misunderstand the setting? Are those settings in kilobytes or something? see_no_evil

There are 3 partitions, btw.

**Comment by npepinpe** [comment]
That's still less than the usage we're seeing. I have a feeling this is misconfigured, sorry for passing on those settings :see_no_evil:

**Comment by deepthidevaki** [comment]
Looks like the memory usage by rocksdb in the cluster with snapshotPeriod = 15m is capped.
```
Total: 1283.7 MB
   576.0  44.9%  44.9%    576.0  44.9% rocksdb::Arena::AllocateNewBlock
   494.7  38.5%  83.4%    494.7  38.5% os::malloc@bb9170
   145.7  11.3%  94.8%    174.5  13.6% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.9%  96.6%     24.0   1.9% std::vector::reserve
    19.1   1.5%  98.1%     23.2   1.8% rocksdb::BlockFetcher::ReadBlockContents
```

While the cluster with snapshotPeriod=1m is still increasing
```
Total: 2487.1 MB
  1764.0  70.9%  70.9%   1764.0  70.9% rocksdb::Arena::AllocateNewBlock
   495.6  19.9%  90.9%    495.6  19.9% os::malloc@bb9170
   145.1   5.8%  96.7%    173.7   7.0% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.0%  97.6%     24.0   1.0% std::vector::reserve
```

**Comment by deepthidevaki** [comment]
Update:

**Broker with snapshotPeriod = 1m**
```
Total: 5623.0 MB
  4843.1  86.1%  86.1%   4843.1  86.1% rocksdb::Arena::AllocateNewBlock
   520.4   9.3%  95.4%    520.4   9.3% os::malloc@bb9170
   145.1   2.6%  98.0%    173.7   3.1% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   0.4%  98.4%     24.0   0.4% std::vector::reserve
```
![image](https://user-images.githubusercontent.com/1997478/98341875-d7ad7700-200f-11eb-8115-0e2b404787d0.png)


**Broker with snapshotPeriod=15m**
```
Total: 1375.8 MB
   654.0  47.5%  47.5%    654.0  47.5% rocksdb::Arena::AllocateNewBlock
   504.3  36.7%  84.2%    504.3  36.7% os::malloc@bb9170
   145.7  10.6%  94.8%    174.5  12.7% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   1.7%  96.5%     24.0   1.7% std::vector::reserve
```
![image](https://user-images.githubusercontent.com/1997478/98341776-b3519a80-200f-11eb-91c6-431ad8b5c7fa.png)

**Broker with writeBufferManager, snapshotPeriod=1m**
```
Total: 2951.5 MB
  2204.0  74.7%  74.7%   2205.5  74.7% rocksdb::Arena::AllocateNewBlock
   519.4  17.6%  92.3%    519.4  17.6% os::malloc@bb9170
   147.5   5.0%  97.3%    175.6   6.0% rocksdb::OptimisticTransactionDB::Open@60a510
    24.0   0.8%  98.1%     24.0   0.8% std::vector::reserve
```
![image](https://user-images.githubusercontent.com/1997478/98341837-c7959780-200f-11eb-9c34-89488d85a074.png)

**Comment by yzhengwei** [comment]
My borker set snapshotPeriod: 15m, but its memory continued to grow until the OOM or I restart it. My  zeebe cluster version is 0.24.3 and it has 3 nodes with 4 core CPU and 32 Memory, How can I fix it?
![image](https://user-images.githubusercontent.com/59439277/100590798-d67e1a00-332f-11eb-8fe2-408786efcea1.png)

**Comment by deepthidevaki** [comment]
@yzhengwei  If you configure rocksdb parameters, it would help to reduce the overall memory consumption. Although we don't have a recommended value for all parameters, setting a lower value for the following might help.

ZEEBE_BROKER_DATA_ROCKSDB_COLUMNFAMILYOPTIONS_WRITE_BUFFER_SIZE  (default = 64MB)
ZEEBE_BROKER_DATA_ROCKSDB_COLUMNFAMILYOPTIONS_MAX_WRITE_BUFFER_SIZE_TO_MAINTAIN (default = 128MB)

Note that the above configuration are available only from Zeebe version 0.25.0

**Comment by rusterholz** [comment]
Using Zeebe 0.25.2, we have set the write_buffer_size to 8MB and the max_write_buffer_size_to_maintain to 16MB as recommended by @npepinpe on the Zeebe Slack. However we saw no difference in the memory consumption patterns of the Zeebe brokers -- memory usage continues to grow and grow until it runs out of memory and is killed by Kubernetes.

This is causing a fair amount of concern in our organization, as we would like to scale up our usage of Zeebe, but we need to be able to keep Zeebe running without worrying that it will exceed whatever memory limits we place upon those containers. Currently it seems that, given enough time, Zeebe will _always_ eventually exceed _any_ memory limit, and therefore there is no way to run a Zeebe broker for an arbitrarily-long time.
